### PR TITLE
fix(console-v2): handle multiline paste and add autocomplete support

### LINKED ISSUE #5983: console-v2 cannot handle paste of multiline statements correctly
### To Reproduce
Copy the statements below, and paste it in [console-v2](https://pyodide.org/en/stable/console-v2.html)

```python
import os
print(os.environ)
```

And then, 

```
print(os.environ)
  File "<console>", line 1
    import os
             ^
SyntaxError: multiple statements found while compiling a single statement

```
I can't only render the multiline statements but execute them.

### Expected behavior
```
>>> import os
... print(os.environ)
...
environ({'USER': ...
```

### Environment

- Pyodide Version: 0.29.0
- Browser version: Chrome 141.0.7390.123

### LINKED ISSUE #5984: console-v2 does not support autocomplete
## 🐛 Bug
Usually, pressing the `TAB` key in a console triggers the autocomplete feature while typing something. But in console-v2, it just inserts spaces instead.

### To Reproduce
```
>>> foobar = 1

# press TAB after typing "foo"
>>> foo

# result ( "^" means position of cursor) 
>>> foo 
        ^

```

### Expected behavior

```
>>> foobar = 1

# press TAB after typing "foo"
>>> foo

# result (expected)
>>> foobar

# Also, pressing tab key twice with "i" should make
>>> i
id(          import       input(       is           issubclass(               
if           in           int(         isinstance(  iter(    


```

### Environment

Pyodide Version: 0.29.0
Browser version: Chrome 141.0.7390.123

### PR DESCRIPTION
**Pull request by rushdarshan**
## Summary

Fixes #5983 - console-v2 cannot handle paste of multiline statements correctly
Fixes #5984 - console-v2 does not support autocomplete

## Changes

- Modified `src/templates/console-v2.html` to detect multiline paste in `term.onData` and execute each line sequentially via `await execLine()`
- Added Tab key handler that calls Python's `complete()` method to display autocomplete suggestions

## Testing

- 10/10 pure Python unit tests pass (`pytest src/tests/test_console.py`)
- 4 node/selenium test errors are pre-existing infrastructure failures requiring a full `dist/` build — unrelated to this change
- Manually tested multiline paste and Tab autocomplete in browser

## Notes

Contribution for FOSS Hack 2026

### CODE REVIEW

**Review by ryanking13 — APPROVED** [review]
Thanks! I didn't read the implementation thoroughly, but tested locally and it works.

@agriyakhetarpal Do you want to review this too?

File: src/templates/console-v2.html [hunk 1]
```diff
@@ -357,16 +357,81 @@
...
+                      );
+                    }
+                  }
+                } else {
+                  // Insert arbitrary string at cursor position
+                  const before = buffer.slice(0, cursorIndex);
+                  const after = buffer.slice(cursorIndex);
+                  const newBuf = before + data + after;
+                  const newCursor = cursorIndex + data.length;
+                  setBuffer(newBuf, newCursor);
```
Comment (agriyakhetarpal) [inline src/templates/console-v2.html]: Looks like this pattern is getting a bit more repeated now. We can make this a helper, such as:

```javascript
function insertAtCursor(text) {
  const before = buffer.slice(0, cursorIndex);
  const after = buffer.slice(cursorIndex);
  setBuffer(before + text + after, cursorIndex + text.length);
}
```

File: src/templates/console-v2.html [hunk 2]
```diff
@@ -357,16 +357,81 @@
...
+                  buffer.slice(0, start) +
+                  completions[0] +
+                  buffer.slice(cursorIndex);
+                setBuffer(newBuf, start + completions[0].length);
+              } else if (completions.length > 1) {
+                const commonPrefix = completions.reduce((a, b) => {
+                  let i = 0;
+                  while (i < a.length && i < b.length && a[i] === b[i])
+                    i++;
+                  return a.slice(0, i);
```
Comment (agriyakhetarpal) [inline src/templates/console-v2.html]: This can be a `longestCommonPrefix` I think:

```js
function longestCommonPrefix(strings) {
  if (strings.length === 0) return "";
  return strings.reduce((a, b) => {
    let i = 0;
    while (i < a.length && i < b.length && a[i] === b[i]) i++;
    return a.slice(0, i);
  });
}
```

File: src/templates/console-v2.html [hunk 3]
```diff
@@ -357,16 +357,81 @@
                 refreshLine();
               }
               break;
+            case "\t": {
+              // Tab key: trigger Python autocomplete via pyconsole.complete()
```
Comment (agriyakhetarpal) [inline src/templates/console-v2.html]: This is doing a lot. I would suggest that you move this into a function:

```js
async function handleTabCompletion() { ... }
```

and then this becomes:

```js
case "\t":
  await handleTabCompletion();
  break;
```

which will be much more readable.

Similarly, you can do this with the multiline pasting in a `async function handlePaste(data) { ... }` function.

File: src/templates/console-v2.html [hunk 4]
```diff
@@ -357,16 +357,81 @@
...
                 data = data.replace(/\u00a0/g, " ");
-                // Insert arbitrary string at cursor position
-                const before = buffer.slice(0, cursorIndex);
-                const after = buffer.slice(cursorIndex);
-                const newBuf = before + data + after;
-                const newCursor = cursorIndex + data.length;
-                setBuffer(newBuf, newCursor);
+                // Detect and handle multiline paste (data contains line breaks)
+                if (data.includes("\r") || data.includes("\n")) {
+                  const lines = data.split(/\r\n|\r|\n/);
```
Comment (agriyakhetarpal) [inline src/templates/console-v2.html]: Is this better?

```suggestion
                const normalized = data.replace(/\r\n?/g, "\n");
                  if (normalized.includes("\n")) {
                  const lines = normalized.split("\n");
```

File: src/templates/console-v2.html [hunk 5]
```diff
@@ -357,16 +357,81 @@
...
+                      const after = buffer.slice(cursorIndex);
+                      setBuffer(
+                        before + line + after,
+                        cursorIndex + line.length,
+                      );
+                      term.write("\r\n");
+                      await execLine(buffer);
+                      buffer = "";
+                      cursorIndex = 0;
+                      term.write(prompt);
```
Comment (agriyakhetarpal) [inline src/templates/console-v2.html]: I think this would be easier to reason about if all buffer/cursor state changes were to go through one helper rather than mixing `setBuffer(...)` with `buffer = ""` / `cursorIndex = 0`. That would also make the paste and completion paths more consistent.

So, to put it in code, I would say:

```js
function clearBuffer() {
  setBuffer("", 0);
}
```

and then you would do

```js
await execLine(buffer);
clearBuffer();
term.write(prompt);
```

Does that make sense?

**Review by agriyakhetarpal — COMMENTED** [review]
I tried out the autocomplete with several scenarios, and it all works well. I found a case of an explosion with `import` statements when testing on https://pyodide--6133.org.readthedocs.build/en/6133/console-v2.html:

```python
import match # or just "import" also
```

gets me


`False  None  True  and   as   assert   async   await   break  class   continue  def   del   elif   else  except   finally:  for   from   global   if   import   in   is   lambda   nonlocal   not   or   pass  raise   return   try:  while   with   yield   _  case   match   type   __name__  __doc__  __package__  __loader__()  __spec__  __annotations__  _pyodide_core  i  x  y  greet()  math  __build_class__(  __import__(  abs(  all(  any(  ascii(  bin(  breakpoint(  callable(  chr(  compile(  delattr(  dir(  divmod(  eval(  exec(  format(  getattr(  globals()  hasattr(  hash(  hex(  id(  input(  isinstance(  issubclass(  iter(  aiter(  len(  locals()  max(  min(  next(  anext(  oct(  ord(  pow(  print(  repr(  round(  setattr(  sorted(  sum(  vars(  Ellipsis  NotImplemented  bool(  memoryview(  bytearray(  bytes(  classmethod(  complex(  dict(  enumerate(  filter(  float(  frozenset(  property(  int(  list(  map(  object()  range(  reversed(  set(  slice(  staticmethod(  str(  super(  tuple(  zip(  __debug__  BaseException(  BaseExceptionGroup(  Exception(  GeneratorExit(  KeyboardInterrupt(  SystemExit(  ArithmeticError(  AssertionError(  AttributeError(  BufferError(  EOFError(  ImportError(  LookupError(  MemoryError(  NameError(  OSError(  ReferenceError(  RuntimeError(  StopAsyncIteration(  StopIteration(  SyntaxError(  SystemError(  TypeError(  ValueError(  Warning(  FloatingPointError(  OverflowError(  ZeroDivisionError(  BytesWarning(  DeprecationWarning(  EncodingWarning(  FutureWarning(  ImportWarning(  PendingDeprecationWarning(  ResourceWarning(  RuntimeWarning(  SyntaxWarning(  UnicodeWarning(  UserWarning(  BlockingIOError(  ChildProcessError(  ConnectionError(  FileExistsError(  FileNotFoundError(  InterruptedError(  IsADirectoryError(  NotADirectoryError(  PermissionError(  ProcessLookupError(  TimeoutError(  IndentationError(  _IncompleteInputError(  IndexError(  KeyError(  ModuleNotFoundError(  NotImplementedError(  PythonFinalizationError(  RecursionError(  UnboundLocalError(  UnicodeError(  BrokenPipeError(  ConnectionAbortedError(  ConnectionRefusedError(  ConnectionResetError(  TabError(  UnicodeDecodeError(  UnicodeEncodeError(  UnicodeTranslateError(  ExceptionGroup(  EnvironmentError(  IOError(  open(  quit(  exit(  copyright()  credits()  license()  help(  ___EvalCodeResultException`

but https://pyodide.org/en/latest/console-v2.html matches what I see on the regular Python (CPython, not Pyodide) console on my machine (because, obviously, there is no autocomplete support on `main` anyway). This is not a big blocker, but just something I observed while testing multiple cases.

The same problem exists for:

```python
import os
```

and then doing

```python
os.path.join(<Tab>
```

With regular Python on my machine, I get nothing. However, with console v2 from this PR, I get the above explosion that spews the Python global namespace dump (i.e., the builtins and keywords).

Could you please investigate and determine whether this is a pre-existing issue surfaced by the changes in this PR, or if it is being caused directly by this PR? I am happy if you would like to put the fix in another PR if you think this would be harder to handle. Thanks!

Next, I have a few inline comments that for improving the code. Please see below:

File: src/templates/console-v2.html (hunk 1 above)
Comment (rushdarshan) [reply src/templates/console-v2.html]: Done – I’ve added an insertAtCursor(text) helper and switched all the places that were doing manual before/after slicing to use it. This makes the buffer updates much easier to follow, thanks for the suggestion.

File: src/templates/console-v2.html (hunk 2 above)
Comment (rushdarshan) [reply src/templates/console-v2.html]: Done – extracted a longestCommonPrefix(strings) helper and use it in the Tab completion path instead of the inline reduce. The implementation matches the version you suggested.

File: src/templates/console-v2.html (hunk 3 above)
Comment (rushdarshan) [reply src/templates/console-v2.html]: Done – I’ve moved all the Tab key logic into an async function handleTabCompletion() and the switch case is now just:
case "\t":
  await handleTabCompletion();
  break;

This made the term.onData handler a lot more readable.

File: src/templates/console-v2.html (hunk 4 above)
Comment (rushdarshan) [reply src/templates/console-v2.html]: Done – paste handling now lives in an async function handlePaste(data), and I first normalise newlines with data.replace(/\r\n?/g, "\n") before splitting on \n. That should behave better across different platforms.

File: src/templates/console-v2.html (hunk 5 above)
Comment (rushdarshan) [reply src/templates/console-v2.html]: Done – added a clearBuffer() helper that just calls setBuffer("", 0), and switched the Enter / Ctrl‑C / multiline paths to use it instead of manually resetting buffer and cursorIndex. That keeps all buffer state changes going through one place.

**Comment by rushdarshan** [comment]
I’ve pushed an update addressing the review feedback:
- Added helpers for inserting at the cursor, clearing the buffer, and computing the longest common prefix.
- Moved Tab completion and paste handling into handleTabCompletion() / handlePaste() for better readability.
- Normalised newline handling for multiline paste and wired all buffer state changes through the helpers.
The autocomplete “explosion” case you observed is still open; right now this PR focuses on the refactor and the original multiline paste + autocomplete support. I’m happy to work on a follow‑up change for the autocomplete behaviour if you think it makes sense to handle that separately.

**Review by agriyakhetarpal — CHANGES_REQUESTED** [review]
> The autocomplete “explosion” case you observed is still open; right now this PR focuses on the refactor and the original multiline paste + autocomplete support. I’m happy to work on a follow‑up change for the autocomplete behaviour if you think it makes sense to handle that separately.

Works for me, thanks!

Testing again on https://pyodide--6133.org.readthedocs.build/en/6133/console-v2.html, I can see another problem here:

```pycon
>>> import math
>>> >>> 
```

The prompt seems to be getting duplicated with each command, and then when I press <kbd>Enter</kbd>, it stays that way:

```pycon
>>> 
>>> >>> 
>>> >>> 
>>> >>> 
>>> >>> 
>>> >>> 
```

It is only after I type something or perform a keystroke in the terminal that the second prompt vanishes, but then it reappears when the output is evaluated from the statement:

```pycon
>>> print("abcd")
abcd
>>> >>> 
>>> >>> 
>>> >>> 
>>> >>> 
>>> >>> 
>>> >>> 
>>> >>> 
>>> >>> 
>>> print(3+5)
8
>>> >>> 
>>> >>> 
>>> >>> 
>>> >>> 
>>> >>> 
```

Could you please investigate?

**Comment by rushdarshan** [comment]
Thanks for catching that prompt duplication! I've investigated the root cause and pushed a couple of commits to address both the prompt issue and the autocomplete explosion you noticed earlier.

Here is a summary of the latest fixes:

- Fixed Prompt Duplication (>>> >>>): The duplicate prompts were being caused by redundant term.write(prompt) calls and implicit line redraws firing sequentially. I've centralized the state changes through the clearBuffer() helper and ensured the prompt is only written once after a block finishes executing.

- Guarded Autocomplete Explosion: Added safeguards to prevent the massive global namespace dump. The handler now skips complete() calls on empty inputs, purely numeric tokens, and right after open delimiters ((, [, {). I also added a cap to the printed completion list so it truncates gracefully rather than flooding the terminal.

- Enhanced Multiline Paste (Transcripts): The paste logic now automatically strips >>>  and ...  prefixes. This means users can directly copy/paste standard Python REPL transcripts and they will execute perfectly without manual cleanup!

- Shift+Tab: Added a case to explicitly ignore Shift+Tab so it no longer accidentally clears the input buffer.

I've manually verified these edge cases in the browser and the unit tests are passing. Let me know how it looks to you now!

**Review by agriyakhetarpal — APPROVED** [review]
Thanks, @rushdarshan! This looks good to me now, and works well! I would recommend adding one more test for the multiline paste behaviour.

 
@ryanking13, would you like to take a second look?

**Comment by rushdarshan** [comment]
Thanks a lot for the review and re-test! I’ve added an additional test for multiline paste behavior as suggested.

**Comment by agriyakhetarpal** [comment]
Looks like the test is failing. You likely need to add another <kbd>Enter</kbd>, to eval the statement and get the output.

File: src/tests/test_console.py [hunk 6]
```diff
@@ -632,6 +664,20 @@ def exec_and_get_result(command):
...
+        term.clear();
+        await new Promise(resolve => setTimeout(resolve, 100));
+        term.paste("import math\\nmath.sqrt(16)\\n");
+        term.paste("\\r");
+        await new Promise(resolve => setTimeout(resolve, 300));
+        """
+    )
+    result = get_terminal_content()
+    assert ">>> >>>" not in result
+    assert "4.0" in result
```
Comment (ryanking13) [inline src/tests/test_console.py]: I guess this should be 4?

Comment (agriyakhetarpal) [reply src/tests/test_console.py]: Indeed, yes, it should be.

Comment (agriyakhetarpal) [reply src/tests/test_console.py]:
```suggestion
    assert "4" in result
```

**Review by agriyakhetarpal — APPROVED** [review]
Thanks for staying put on this, @rushdarshan! Let's merge once tests pass.

**Comment by agriyakhetarpal** [comment]
Looks like the RTD build is failing, but I don't see that on other PRs. Probably some sort of flake? I'll trigger CI again...

**Comment by agriyakhetarpal** [comment]
All good this time – merging!

**Comment by gpotter2** [comment]
Hi guys, just bumped into this PR. FYI I had made a wrapper for our website that uses pyodide with xterm.js, pretty similar to what this PR is trying to do. https://scapy.net/?try=1 (code [here](https://github.com/secdev/secdev.github.io/blob/master/src/components/XTermScapy.vue)). It's in typescript, and has a slightly different set of features compared to console-v2, like it has color formatting, Ctrl^C/Ctrl^L/Ctrl+arrows support, but the history isn't saved in the browser and it has no completion.

image

I think it would be cool to aim at putting the REPL code into a separate project that other websites can use as a dependency. I think it's a bit sad that everyone has to reimplement this themselves (you guys with console-v2, myself, the guy from https://whitefirer.org/html/pyodide/, etc.) and I believe it's a cool demo for pyodide if many websites could easily embed it.

If you guys are interested in making this a standalone repo, and merging the projects, feel free to reach out. Feel free to take a look at the code

**Comment by ryanking13** [comment]
@gpotter2 Thanks for the suggestion. Well, not sure what other maintainers think, but I personally don't want to put too much effort on the repl. We intentionally keep this repl small and compact so that it only serves as an example.
